torchmultimodal/utils/attention.py

```diff
@@ -0,0 +1,33 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from torch import Tensor
+
+
+def get_extended_attention_mask(attention_mask: Tensor) -> Tensor:
+    """
+    Makes broadcastable attention and causal masks so that future and masked tokens are ignored.
+
+    Args:
+        attention_mask (Tensor): Mask with ones indicating tokens to attend to, zeros for tokens to ignore.
+    Returns:
+        extended_attention_mask (Tensor): extended attention mask with the same dtype as attention_mask.dtype.
+    """
+
+    if attention_mask.dim() == 3:
+        extended_attention_mask = attention_mask[:, None, :, :]
+    elif attention_mask.dim() == 2:
+        extended_attention_mask = attention_mask[:, None, None, :]
+    else:
+        raise ValueError(
+            "Wrong shape for attention_mask (shape {})".format(attention_mask.shape)
+        )
+
+    extended_attention_mask = extended_attention_mask.to(
+        dtype=attention_mask.dtype
+    )  # fp16 compatibility
```

Comment:
Do you know why after adding `None` the data type might change (such that here it's forced back to that of the input attn mask)? 

Is it related to the comment "fp16 compatibility"?

Reply:
I am not sure why this is needed, the default data type for torch tensors is `float32`. I supposed if you passed in an attention mask as fp16 this would preserve it, but throughout the code base I see attention_mask be instantiated with `float32` ([here](https://github.com/facebookresearch/multimodal/blob/df2f952d7751bf6517ee85686802a72a8b5d432e/torchmultimodal/models/flava/flava_text_encoder.py#L180) or even an `IntTensor` ([here](https://github.com/facebookresearch/multimodal/blob/df2f952d7751bf6517ee85686802a72a8b5d432e/test/models/flava/test_flava_text_encoder.py#L79)). 

There is also the `Fp32LayerNorm` in the normalization folder, which I'm not sure is really doing anything different than `nn.LayerNorm` since weights and biases should already be `float32` by default.

Reply:
probably needed for mixed precision training

Reply:
If the idea is to enforce fp16 for mixed precision training, then should we set `dtype` equal to `fp16` instead of inheriting that of the input attention_mask?